content/en/blog/_posts/2023-01-12-protect-mission-critical-pods-priorityclass/index.md

```diff
@@ -0,0 +1,300 @@
+  - In the background, the scheduler runs as an infinite loop looking for pods without a `nodeName` set
+    (and that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)).
+  - For each Pod that needs scheduling, the scheduler tries to make its best decision about which node should run that Pod.
+  - To name a few, `nodeSelector` , `taints and tolerations` , `nodeAffinity` , the rank of nodes based on available resources (for example, CPU and memory), and several other criteria are used to determine the pod's placement.  
+- Just in case the scheduler is not able to find any node, the pod remains to be in the pending state, which is not an ideal scenario.
+
+The below diagram explains how the request flows in (starting from point number 1 to 4):
+
+{{< figure src=kube-scheduler.svg alt="A diagram showing the scheduling of three Pods that a client has directly created." title="Scheduling in Kubernetes">}}
+
+## Typical use cases
+
+Below are some real-life scenarios where control over the scheduling and eviction of pods may be required.
+
+1. Let's say the pod you are planning to deploy is really critical, and you have some resource constraints. How will you make the resources available in the cluster? by first identifying which pods are not required, then manually deleting them, etc., which is a manual and tedious process, or you may plan to add a new node to the cluster, which too might take some time and money. 
+
+   An example of such a case is a DaemonSet of an infrastructure component like Grafana Loki. In that case, the Loki pods must run on each node before any other pods run on that node.
+
+2. Another use case can be: Suppose, you have a single cluster, and that cluster holds the pods for the below environments:
+   - preprod [Priority 1] 
+   - staging [Priority 2] 
+   - development [Priority 3]
+
+   Unfortunately, the cluster goes through high resource consumption, and now there is a resource competition on the nodes with respect to the CPU and the memory. At least in the short term, you don’t have further nodes to expand or scale that Kubernetes cluster (cluster level autoscaling _may_ add more nodes, but this takes time). Perhaps some pods will remain in a Pending state, or perhaps service will degrade as Pods compete for resources. If the kubelet does evict a Pod from a node, that eviction happens at random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.
+
+3. Assume you have a microservices architecture that includes Kafka, Zookeeper, MongoDB, a couple of jobs, a couple of CronJobs, and around 40+ microservices. Assume that if we are caught up in this situation of resource crunch and somehow Kafka or MongoDB gets evicted, then all our microservices will be useless until and unless these pods are up and start serving traffic again. 
```

Comment:
This is a very specific microservice architecture. I'd recommend that we provide a generalized one and more context on why all microservices would be rendered useless if the Kafka/MongoDB pods get evicted. I know it sounds like I'm dumbing down things here, but readers may not have the context we have.

Reply:
Does this sounds good?

Assume you have a cloud-native application, i.e., a collection of small, independent, loosely coupled services and a couple of stateful applications like a queuing service or a database. And assume that if we are caught up in this situation of resource crunch and somehow the database pod gets evicted, then all our small services will be useless until and unless the database pod is up and start to serve traffic again.

Reply:
Just rephrased your suggestion a bit.
```suggestion
3. A third example could be a microservice backed by a queuing application or a database running into a resource crunch and the queue or database getting evicted. In such a case, all the other services would be rendered useless until the database can serve traffic again.
```

Reply:
Looks good, Updated.